sentry-android-replay/src/main/java/io/sentry/android/replay/viewhierarchy/ViewHierarchyNode.kt

```diff
@@ -0,0 +1,173 @@
+                    nodeInfo.recycle()
+                    if (shouldRedact) {
+                        val bounds = Rect()
+                        val text = view.text.toString()
+                        view.paint.getTextBounds(text, 0, text.length, bounds)
+                        dominantColor = adjustAlpha(view.currentTextColor)
+                        rect = Rect()
+                        view.getGlobalVisibleRect(rect)
+
+                        var textEnd = Int.MIN_VALUE
+                        var textStart = Int.MAX_VALUE
+                        if (view.layout != null) {
+                            for (i in 0 until view.layout.lineCount) {
+                                val min = view.layout.getLineStart(i)
+                                val minPosition = view.layout.getPrimaryHorizontal(min).toInt()
+                                val max = view.layout.getLineVisibleEnd(i)
+                                val maxPosition = view.layout.getPrimaryHorizontal(max).toInt()
+                                if (minPosition < textStart) {
+                                    textStart = minPosition
+                                }
+                                if (maxPosition > textEnd) {
+                                    textEnd = maxPosition
+                                }
+                            }
+                        } else {
+                            textEnd = rect.right - rect.left
+                            textStart = 0
+                        }
+                        // TODO: support known 3rd-party widgets like MaterialButton with an icon
+                        // TODO: also calculate height properly based on text bounds
+                        rect.left += textStart + view.paddingStart
+                        rect.right = rect.left + (textEnd - textStart)
+                    }
+                }
+
+                is ImageView -> {
+                    shouldRedact = isVisible(view) && (view.drawable?.isRedactable() ?: false)
+                    if (shouldRedact) {
+                        dominantColor = adjustAlpha((view.drawable?.pickDominantColor() ?: Color.BLACK))
+                        rect = Rect()
+                        view.getGlobalVisibleRect(rect)
+                    }
+                }
+            }
+            return ViewHierarchyNode(
+                view.x,
+                view.y,
+                view.width,
+                view.height,
+                shouldRedact,
+                dominantColor,
+                rect
+            )
+        }
+
+        private fun Drawable.isRedactable(): Boolean {
+            return when (this) {
+                is InsetDrawable, is ColorDrawable, is VectorDrawable, is GradientDrawable -> false
+                is BitmapDrawable -> !bitmap.isRecycled && bitmap.height > 0 && bitmap.width > 0
+                else -> true
+            }
+        }
+
+        private fun Drawable.pickDominantColor(): Int {
```

Comment:
m: This sounds heavy. Creating a new bitmap to calculate the average color seems to be an overkill, can we just use black or white based on the theme?

Reply:
yeah, it doesn't look as nice with black rectangles, but I also left a TODO down below sayin the same thing - maybe we can expose an option for that, and default to black/white rectangle. I'll also look into doing it off the main thread if possible

Reply:
We've improved this significantly in https://github.com/getsentry/sentry-java/pull/3203/commits/1ce57cbd24afa08a583d28e0ab6e2365a0592203 - now we just reuse the same single-pixel bitmap where we draw portions of the screenshot that need redacting + also doing this off-the-main-thread